A33-Fault-Injection.md

```diff
@@ -0,0 +1,262 @@
+
+## Alternatives
+
+### Alternative 1: Fault Injection Config via ProtoBuf Metadata
+
+After ConfigSeletor calculated the final fault injection config, it can encode the config proto and set it as metadata for that RPC. This approach is straightforward and allows more cohesive logic in the fault injection enforcer. On the other hand, it requires additional serialization and deserialization of proto messages, and it can reduce debuggability.
+
+```bash
+grpc-fault-injection-policy-bin=<...>  # The encoded FI config proto message
+```
+
+However, setting additional metadata may leak implementation details onto wire, which is undesired.
+
+### Alternative 2: Fault Injection Config via Multiple Metadata
+
+Fault injection config is part of RDS. However, if we want to support fault injection in non-xDS cases, we have to have a more generic expression of FI config. This approach uses metadata as the final fault injection options for each RPC:
+
+```bash
+grpc-fault-injection-delay-ms=1.5       # 1.5ms delay before sending first request
+grpc-fault-injection-delay-percent=0.2  # This call has 20% chance to be delayed
+grpc-fault-injection-abort-code=7       # Reject the call as PERMISSION_DENIED
+grpc-fault-injection-abort-percent=0.5  # This call has 50% chance to be rejected
+```
+
+In this way, users can utilize fault injection without configuring xDS. It also allows decouple the enforcement of FI from xDS control path.
+
+But similar to above approach, adding extra metadata is not a good idea.
+
+### Alternative 3: Fault Injection Config via ServiceConfig
+
+Across Core, Java and Go, Service Config is a mechanism to supply advanced configuration for specific service or method. It’s natural to add a new section of fault injection policy into the MethodConfig. The proposed proto configuration looks like:
+
+```protobuf
+message MethodConfig {
+
+ // ...ignoring other fields
+
+ // The fault injection policy for outbound RPCs.
+ //
+ // The fault injection offers delay injection and aborting RPC functionality.
+ // Both of them can be specified at the same time. If both features are
+ // triggered, the RPC will be delayed first then abort with the status code.
+ message FaultInjectionPolicy {
+   // The fixed delay before starting the RPC.
+   //
+   // No traffic should be sent until the delay phase ends. By default, the
+   // delay duration is 0 (0 seconds and 0 nanos). If the delay is 0, the RPC
+   // delay feature is disabled and delay_fraction will be ignored.
+   google.protobuf.Duration delay = 1;
+
+   // The possibility that this RPC will be delayed.
```

Comment:
Should it say probability?

Reply:
You are right, probability is more accurate here. Updated.

Reply:
Even probability doesn't quite sound right in this context. What we are really talking about is, for example, out of 1 million RPCs delay will be introduced for these many RPCs (`delay_fraction`). If this is done on a random basis then probability applies to a given RPC being delayed because of this feature. However the number being described here is the "fraction" of RPCs (out of a million) that will be delayed by this feature.